filenamify.js

```diff
@@ -21,14 +21,65 @@ const filenamify = (string, options = {}) => {
 		throw new Error('Replacement string cannot contain reserved filename characters');
 	}
 
+	const getReplacementIndexSet = (string, replacement) => {
+		const indexList = [];
+		if (replacement.length === 0) {
+			return indexList;
+		}
+
+		let position = 0;
+		while (position < string.length) {
+			const index = string.indexOf(replacement, position);
+			if (index >= 0) {
+				indexList.push(index);
+				position = index + replacement.length;
+			} else {
+				break;
+			}
+		}
+
+		return new Set(indexList);
+	};
+
+	const preIndexSet = getReplacementIndexSet(string, replacement);
+	const incrementalReplacementSet = [];
+	for (const incre of [filenameReservedRegex(), reControlChars, reRelativePath]) {
+		for (const iterator of string.matchAll(incre)) {
+			incrementalReplacementSet.push(iterator.index);
+		}
+	}
+
+	const adjustedPreIndexSet = [...preIndexSet];
+	for (const incIndex of incrementalReplacementSet) {
+		for (let index = 0; index < adjustedPreIndexSet.length; index++) {
+			if (adjustedPreIndexSet[index] > incIndex) {
+				adjustedPreIndexSet[index] += replacement.length - 1;
+			}
+		}
+	}
```

Comment:
If possible, could you try to move some of functionality here out into top-level functions? Same with the stuff on line 67.